// vminfod/zonewatcher.js
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at http://smartos.org/CDDL
 *
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file.
 *
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 *
 * Copyright 2015, Joyent, Inc.
 *
 */

var assert = require('assert');
var EventEmitter = require('events').EventEmitter;
var util = require('util');

var SyseventStream = require('/usr/vm/node_modules/sysevent-stream');

module.exports = ZoneWatcher;
module.exports.ZoneWatcher = ZoneWatcher;
function ZoneWatcher(options) {
    var self = this;

    self.options = options;
    assert(options.log, 'no logger passed to ZoneWatcher()');
    self.log = options.log;

    // create a Sysevent stream
    var opts = {
        logger: self.log,
        class: 'status',
        channel: 'com.sun:zones:status'
    };

    self.se = new SyseventStream(opts);
    self.se.on('readable', function () {
        var ev;
        while ((ev = self.se.read()) !== null) {
            var o = {
                type: 'change',
                zonename: ev.data.zonename,
                newstate: ev.data.newstate,
                time: ev.time,
                ev: ev
            };

            self.emit('change', o);
            self.emit('all', o);
        }
    });
}
util.inherits(ZoneWatcher, EventEmitter);

ZoneWatcher.prototype.shutdown = function ()
{
    this.se.stop();
};
