// vmevent/queue.js
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at http://smartos.org/CDDL
 *
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file.
 *
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 *
 * Copyright (c) 2014, Joyent, Inc. All rights reserved.
 *
 */

var async = require('/usr/node/node_modules/async');

exports = module.exports = Queue;

function Queue(opts) {
    if (!opts) {
        opts = {};
    }

    // hash map to contain channel->job mapping
    this.channels = {};

    // the number of concurrent jobs per channel
    this.workers = (opts.workers === undefined) ? 1 : opts.workers;

    // lock to prevent workers from running
    this.paused = (opts.paused === undefined) ? false : opts.paused;

    // array of callbacks once all channels have confirmed pause
    this.pause_callbacks = [];

    // a timestamp
    this.stop_at;

    // array of callbacks once all channels have confirmed fast forwarded
    this.ff_callbacks = [];

    return (this);
}

Queue.prototype.enqueue = function (id, job, callback) {
    var channel;
    var self = this;
    var key;


    if (callback === undefined) {
        callback = function () {};
    }

    if (this.channels[id] === undefined) {
        this.channels[id] = {
            jobs: {},
            working: 0
        };
    }

    channel = this.channels[id];
    key = (new Date()).toISOString();

    channel.jobs[key] = {
        work: job,
        callback: callback
    };

    process.nextTick(function () {
        self.work(id);
    });
}

Queue.prototype.work = function (id) {
    var self = this;
    var channel;
    var key;
    var job;

    // queue is paused and we're not fast-forwarding
    if (this.paused === true && this.stop_at === undefined) {
        return;
    }

    // channel is empty
    if (this.channels[id] === undefined) {
        return;
    }

    channel = this.channels[id];

    // workers are busy
    if (channel.working === this.workers) {
        return;
    }

    // since the keys are a timestamp, sort the keys and return
    // the first item to get the oldest.
    key = Object.keys(channel.jobs).sort()[0];

    // we've fast-forwarded
    if (this.stop_at !== undefined) {
        if (key > this.stop_at) {
            return;
        }
    }

    job = channel.jobs[key];

    // increment the working count
    channel.working++;

    job.work(function () {
        // decrement working count
        channel.working--;

        // run the original callback, passing along the arguments
        job.callback.apply(null, arguments);

        // delete the job
        delete (channel.jobs)[key];

        // checkPaused if nothing is working
        if (channel.working === 0) {
            self._checkPaused();
        }

        // checkFastForwarded if nothing is working
        if (channel.working === 0) {
            self._checkFastForwarded();
        }

        // cleanup the channel if there are no more jobs
        if (Object.keys(channel.jobs).length === 0) {
            delete (self.channels)[id];
        }

        process.nextTick(function () {
            self.work(id);
        });
    });
}

Queue.prototype.pause = function (callback) {
    this.paused = true;
    this.pause_callbacks.push(callback);
    this._checkPaused();
}

Queue.prototype.resume = function () {
    this.paused = false;
    this.stop_at = undefined;
    this.pause_callbacks = [];
    this.ff_callbacks = [];

    async.each(Object.keys(this.channels), function (id) {
        this.work(id);
    });
}

Queue.prototype.fastForward = function (timestamp, callback) {
    if (this.paused !== true) {
        callback(new Error('cannot fastForward a queue that is not paused'));
        return;
    }

    this.stop_at = timestamp;
    this.ff_callbacks.push(callback);
    this._checkFastForwarded();
}

Queue.prototype._checkPaused = function () {
    if (this.paused !== true) {
        return;
    }

    for (key in this.channels) {
        if (this.channels[key].working > 0) {
            // something is still working
            return;
        }
    }

    this.pause_callbacks.forEach(function (cb) {
        cb();
    });
}

Queue.prototype._checkFastForwarded = function () {
    if (this.stop_at === undefined) {
        return;
    }

    for (id in this.channels) {
        if (this.channels[id].working > 0) {
            // something is still working
            return;
        }
        for (key in this.channels[id].jobs) {
            if (key < this.stop_at) {
                // something still needs to work
                return;
            }
        }
    }

    this.ff_callbacks.forEach(function (cb) {
        cb();
    });
}
