// cloudinit/lofs-fat16.js
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at http://smartos.org/CDDL
 *
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file.
 *
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 *
 * Copyright 2026 Edgecast Cloud LLC.
 */

/*
 * FAT16 Filesystem Image Builder using Native SmartOS Tools
 *
 * Creates a FAT16 filesystem on a zvol using lofiadm, mkfs_pcfs, and mount.
 *
 */

var fs = require('fs');
var path = require('path');
var child_process = require('child_process');
var exec = child_process.exec;

var assert = require('/usr/node/node_modules/assert-plus');

var vasync = require('/usr/vm/node_modules/vasync');
var pipeline = vasync.pipeline;
var forEachPipeline = vasync.forEachPipeline;

/*
 * Create a FAT16 filesystem on the specified zvol and write cloud-init files.
 *
 * @param {String} zvolPath - Path to raw zvol device (e.g. /dev/zvol/rdsk/...)
 * @param {Object} files - Cloud-init files to write, where the key is the
 *                         desired filename and the value is a string of data
 *                         to write to the file.
 * @param {Function} callback - Callback(err)
 */
function createImage(zvolPath, files, callback) {
    assert.string(zvolPath, 'zvolPath');
    assert.object(files, 'files');
    assert.func(callback, 'callback');

    var pipelineArgs = {
        zvolPath: zvolPath,
        lofiDevice: null,
        lofiRawDevice: null,
        mountPoint: null,
        mounted: false,
        files: []
    };

    // Validate filenames and build list of files to write
    for (var fname in files) {
        if (!files.hasOwnProperty(fname)) {
            continue;
        }

        // Skip undefined/null, but include empty strings (0-byte files)
        if (typeof (files[fname]) === 'string') {
            pipelineArgs.files.push({
                name: fname,
                content: files[fname]
            });
        }
    }

    pipeline({
        arg: pipelineArgs,
        funcs: [
            function createMountPoint(ctx, next) {
                exec('mktemp -d /tmp/lofs-fat16-XXXXXX',
                    function _mktemp(err, stdout) {
                    if (err) {
                        next(new Error('mktemp failed to create mount point: '
                            + err.message));
                        return;
                    }
                    ctx.mountPoint = stdout.trim();
                    next();
                });
            },
            function attachLofi(ctx, next) {
                exec('lofiadm -a ' + ctx.zvolPath,
                    function _lofiadm(err, stdout, stderr) {
                    if (err) {
                        next(new Error('lofiadm -a failed: '
                            + (stderr || err.message)));
                        return;
                    }

                    ctx.lofiDevice = stdout.trim();

                    // Construct raw device path for mkfs
                    // /dev/lofi/1 -> /dev/rlofi/1
                    ctx.lofiRawDevice = ctx.lofiDevice
                        .replace('/lofi/', '/rlofi/');

                    next();
                });
            },

            function formatFilesystem(ctx, next) {
                // size of 32768 = 16MiB using 512 byte sectors
                var mkfsOpts = 'b=CIDATA,nofdisk,fat=16,size=32768';
                var mkfsCmd = 'yes | mkfs -F pcfs -o '
                    + mkfsOpts + ' ' + ctx.lofiRawDevice;

                exec(mkfsCmd, function _mkfs(err, stdout, stderr) {
                    if (err) {
                        next(new Error('mkfs_pcfs failed: '
                            + (stderr || err.message)));
                        return;
                    }
                    next();
                });
            },

            function mountFilesystem(ctx, next) {
                var mountCmd = 'mount -F pcfs -o rw ' + ctx.lofiDevice
                    + ' ' + ctx.mountPoint;

                exec(mountCmd, function _mount(err, stdout, stderr) {
                    if (err) {
                        next(new Error('mount failed: '
                            + (stderr || err.message)));
                        return;
                    }
                    ctx.mounted = true;
                    next();
                });
            },

            function writeFiles(ctx, next) {
                forEachPipeline({
                    inputs: ctx.files,
                    func: function _writeFile(fileInfo, cb) {
                        var filepath = path.join(ctx.mountPoint, fileInfo.name);
                        fs.writeFile(filepath, fileInfo.content, 'utf8', cb);
                    }
                }, function writeFilesPipeline(err) {
                    if (err) {
                        next(new Error('Failed to write file: ' + err.message));
                        return;
                    }
                    next();
                });
            },

            // unmount filesystem
            function unmountFilesystem(ctx, next) {
                exec('umount ' + ctx.mountPoint,
                    function _umount(err, stdout, stderr) {
                    if (err) {
                        next(new Error('umount failed: '
                            + (stderr || err.message)));
                        return;
                    }
                    ctx.mounted = false;
                    next();
                });
            },

            // Remove mount point
            function removeMountPoint(ctx, next) {
                fs.rmdir(ctx.mountPoint, function _rmdir(err) {
                    if (err) {
                        next(new Error('rmdir ' + ctx.mountPoint + ' failed: '
                            + err.message));
                        return;
                    }
                    ctx.mountPoint = null;
                    next();
                });
            },

            // Detach lofi device
            function detachLofi(ctx, next) {
                exec('lofiadm -d ' + ctx.lofiDevice,
                    function _lofiadm(err, stdout, stderr) {
                    if (err) {
                        next(new Error('lofiadm -d failed: '
                            + (stderr || err.message)));
                        return;
                    }
                    ctx.lofiDevice = null;
                    next();
                });
            }
        ]
    },
    function pipelineDone(pipelineErr) {
        if (!pipelineErr) {
            callback();
            return;
        }

        // Run cleanup, then return original error
        pipeline({
            arg: pipelineArgs,
            funcs: [

                function unmountIfMounted(ctx, next) {
                    if (!ctx.mounted) {
                        next();
                        return;
                    }
                    exec('umount ' + ctx.mountPoint, function _umount(err) {
                        // Ignore unmount errors during cleanup
                        ctx.mounted = false;
                        next();
                    });
                },

                function rmMountPoint(ctx, next) {
                    if (!ctx.mountPoint) {
                        next();
                        return;
                    }
                    fs.rmdir(ctx.mountPoint, function _rmdir(err) {
                        // Ignore rmdir errors during cleanup
                        ctx.mountPoint = null;
                        next();
                    });
                },

                function lofiDetach(ctx, next) {
                    if (!ctx.lofiDevice) {
                        next();
                        return;
                    }
                    exec('lofiadm -d ' + ctx.lofiDevice,
                        function _lofiadm(err) {
                        // Ignore detach errors during cleanup
                        ctx.lofiDevice = null;
                        next();
                    });
                }
            ]
        },
        function cleanupPipeline() {
            callback(pipelineErr);
        });
    });
}

module.exports = {
    createImage: createImage
};
