// cloudinit/nocloud.js
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at http://smartos.org/CDDL
 *
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file.
 *
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 *
 * Copyright 2026 Edgecast Cloud LLC.
 */

/*
 * Creates a volume containing meta-data, network-config, user-data, and
 * vendor-data cloudinit files for the NoCloud datasource.
 */

var assert = require('/usr/node/node_modules/assert-plus');
var ipUtil = require('../ip');
var lofs = require('./lofs-fat16');

/*
 * Build cloud-init network-config (version 2) from payload.add_nics.
 *
 * Converts SmartOS NIC configuration to cloud-init network-config format.
 */
function networkConfig(payload) {
    assert.object(payload, 'payload');

    // Check for customer-provided network-config first
    var customerMdata = payload.set_customer_metadata
                        || payload.customer_metadata || {};
    if (customerMdata.hasOwnProperty('cloud-init:network-config')) {
        return customerMdata['cloud-init:network-config'];
    }

    var nics = payload.add_nics || [];

    // No network-config if no NICs
    if (nics.length === 0) {
        return undefined;
    }

    var resolvers = payload.resolvers || [];
    var ethernets = {};

    nics.forEach(function (nic) {
        var ifaceName = nic.interface || 'eth0';
        var ethConfig = {};

        // Match by MAC address for reliable interface identification
        if (nic.mac) {
            ethConfig.match = {
                macaddress: nic.mac.toLowerCase()
            };
            // Set the interface name explicitly
            ethConfig['set-name'] = ifaceName;
        }

        // Handle IP addresses
        if (nic.ips && nic.ips.length > 0) {
            // Check for DHCP (IPv4) and addrconf (IPv6 SLAAC)
            var hasDhcp = nic.ips.some(function (ip) {
                return ip === 'dhcp';
            });
            var hasAddrconf = nic.ips.some(function (ip) {
                return ip === 'addrconf';
            });

            if (hasDhcp) {
                ethConfig.dhcp4 = true;
            }
            if (hasAddrconf) {
                ethConfig.dhcp6 = true;
            }

            // Static IPs - ips array contains CIDR notation
            var staticIps = nic.ips.filter(function (ip) {
                return ip !== 'dhcp' && ip !== 'addrconf';
            });
            if (staticIps.length > 0) {
                ethConfig.addresses = staticIps;
            }
        } else if (nic.ip && nic.netmask) {
            // Legacy format: separate ip and netmask fields
            var cidr = ipUtil.netmaskToBits(nic.netmask);
            ethConfig.addresses = [nic.ip + '/' + cidr];
        }

        // Only set the default gateway on the primary NIC.
        if (nic.primary) {
            if (nic.gateways && nic.gateways.length > 0) {
                // cloud-init v2 uses 'gateway4' for IPv4 default gateway
                nic.gateways.forEach(function (gw) {
                    if (gw.indexOf(':') === -1) {
                        // IPv4 gateway
                        ethConfig.gateway4 = gw;
                    } else {
                        // IPv6 gateway
                        ethConfig.gateway6 = gw;
                    }
                });
            } else if (nic.gateway) {
                // Legacy single gateway field
                if (nic.gateway.indexOf(':') === -1) {
                    ethConfig.gateway4 = nic.gateway;
                } else {
                    ethConfig.gateway6 = nic.gateway;
                }
            }
        }

        // Add nameservers to primary NIC
        if (nic.primary && resolvers.length > 0) {
            ethConfig.nameservers = {
                addresses: resolvers
            };
        }

        ethernets[ifaceName] = ethConfig;
    });

    // Build the YAML manually (no YAML library in Node 0.10)
    // Note: network-config file does NOT have a top-level 'network:' key
    var lines = [];
    lines.push('version: 2');
    lines.push('ethernets:');

    Object.keys(ethernets).forEach(function (ifaceName) {
        var eth = ethernets[ifaceName];
        lines.push('  ' + ifaceName + ':');

        if (eth.match) {
            lines.push('    match:');
            lines.push('      macaddress: "' + eth.match.macaddress + '"');
        }

        if (eth['set-name']) {
            lines.push('    set-name: ' + eth['set-name']);
        }

        if (eth.dhcp4) {
            lines.push('    dhcp4: true');
        }

        if (eth.dhcp6) {
            lines.push('    dhcp6: true');
        }

        if (eth.addresses && eth.addresses.length > 0) {
            lines.push('    addresses:');
            eth.addresses.forEach(function (addr) {
                lines.push('      - ' + addr);
            });
        }

        if (eth.gateway4) {
            lines.push('    gateway4: ' + eth.gateway4);
        }

        if (eth.gateway6) {
            lines.push('    gateway6: ' + eth.gateway6);
        }

        if (eth.nameservers && eth.nameservers.addresses) {
            lines.push('    nameservers:');
            lines.push('      addresses:');
            eth.nameservers.addresses.forEach(function (ns) {
                lines.push('        - ' + ns);
            });
        }
    });

    return lines.join('\n') + '\n';
}

function metaDataConfig(payload) {
    assert.object(payload, 'payload');

    // Build meta-data content
    var hostname = payload.hostname || payload.alias || payload.uuid;
    var metaData = '# cloud-init meta-data\n';
    metaData += 'instance-id: ' + payload.uuid + '\n';
    metaData += 'local-hostname: ' + hostname + '\n';
    return metaData;
}

function userDataConfig(payload) {
    assert.object(payload, 'payload');

    var customerMdata = payload.set_customer_metadata
                        || payload.customer_metadata || {};
    if (customerMdata.hasOwnProperty('cloud-init:user-data')) {
        return customerMdata['cloud-init:user-data'];
    }

    /*
     * Always include at least a minimal #cloud-config even if customer doesn't
     * provide one. This is a workaround for at least one nocloud implementation
     * (FreeBSD's nuageinit) which expects user-data exist and be missing,
     * skipping network-config processing.
     * This is a nuageinit quirk, not a cloud-init/nocloud spec requirement.
     */
    return '#cloud-config\n{}\n';
}

function vendorDataConfig(payload) {
    var customerMdata = payload.set_customer_metadata
                        || payload.customer_metadata || {};
    if (customerMdata.hasOwnProperty('cloud-init:vendor-data')) {
        return customerMdata['cloud-init:vendor-data'];
    }
    return undefined;
}

/*
 * Add a nocloud disk entry to the payload and adjust quota or
 * flexible_disk_size to account for the additional space.
 *
 * This is called during VM creation (before volumes are created) to
 * insert the nocloud CIDATA zvol into the disk list.
 */
function updatePayloadDisks(log, payload) {
    assert.object(log, 'log');
    assert.object(payload, 'payload');

    var diskSizeMiB = lofs.DISK_SIZE_MIB;
    var diskSizeGiB = Math.ceil(diskSizeMiB / 1024);

    payload.add_disks.push({
        size: diskSizeMiB,
        model: 'virtio',
        boot: false,
        media: 'disk',
        block_size: lofs.BLOCK_SIZE,

        // This property is used to identify the disk later when
        // adding the cloudinit YAML files to the disk. It is not
        // saved to the final machine JSON object.
        cloudinit_datasource: 'nocloud'
    });

    if (payload.hasOwnProperty('flexible_disk_size')) {
        var oldFlexSize = payload.flexible_disk_size;
        payload.flexible_disk_size = oldFlexSize + diskSizeMiB;

        log.info({
            oldFlexibleDiskSize: oldFlexSize,
            additionalSize: diskSizeMiB,
            newFlexibleDiskSize: payload.flexible_disk_size
        }, 'Increased flexible_disk_size for nocloud disk');
    } else {
        var oldQuotaGiB = payload.quota || 0;
        if (oldQuotaGiB < diskSizeGiB) {
            payload.quota = oldQuotaGiB + diskSizeGiB;
            log.info({
                oldQuotaGiB: oldQuotaGiB,
                additionalQuotaGiB: diskSizeGiB,
                newQuotaGiB: payload.quota
            }, 'Increased quota for nocloud disk');
        }
    }
}

/*
 * Validate that the disks array contains a nocloud CIDATA zvol at the
 * given path with the expected size, block_size, and boot properties.
 *
 * Returns true if a matching disk is found, false otherwise.
 */
function validateNoCloudDisk(disks, path) {
    assert.arrayOfObject(disks, 'disks');
    assert.string(path, 'path');

    for (var i = 0; i < disks.length; i++) {
        if (disks[i].path === path
            && disks[i].size === lofs.DISK_SIZE_MIB
            && disks[i].block_size === lofs.BLOCK_SIZE
            && disks[i].boot === false) {
            return true;
        }
    }

    return false;
}

function createPCFS(log, payload, path, callback) {
    assert.object(log, 'log');
    assert.object(payload, 'payload');
    assert.string(path, 'path');
    assert.func(callback, 'callback');

    var files = {
        'meta-data': metaDataConfig(payload),
        'network-config': networkConfig(payload)
    };

    var user = userDataConfig(payload);
    if (user) {
        files['user-data'] = user;
    }

    var vendor = vendorDataConfig(payload);
    if (vendor) {
        files['vendor-data'] = vendor;
    }

    // Create CIDATA image and write to path
    lofs.createImage(path, files, function (err) {
        if (err) {
            log.error({err: err}, 'Failed to create cloudinit CIDATA image');
            callback(err);
            return;
        }

        log.info({
            path: path
        }, 'Successfully wrote cloudinit CIDATA image to zvol');
        callback();
    });
}

module.exports = {
    createPCFS: createPCFS,
    updatePayloadDisks: updatePayloadDisks,
    validateNoCloudDisk: validateNoCloudDisk,
    _metaDataConfig: metaDataConfig,
    _networkConfig: networkConfig,
    _userDataConfig: userDataConfig,
    _vendorDataConfig: vendorDataConfig
};
